### PR TITLE
add v prediction

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds v-prediction for DDIM and Euler scheduler.

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 1]
```diff
@@ -258,7 +261,19 @@ def step(
 
         # 3. compute predicted original sample from predicted noise also called
         # "predicted x_0" of formula (12) from https://arxiv.org/pdf/2010.02502.pdf
-        pred_original_sample = (sample - beta_prod_t ** (0.5) * model_output) / alpha_prod_t ** (0.5)
+        if self.prediction_type == "epsilon":
+            pred_original_sample = (sample - beta_prod_t ** (0.5) * model_output) / alpha_prod_t ** (0.5)
+            eps = torch.tensor(1)
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:266]: Why do we need that?

Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:266]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -258,7 +261,19 @@ def step(
...
+        elif self.prediction_type == "sample":
+            pred_original_sample = model_output
+            eps = torch.tensor(1)
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:269]:
```suggestion
            eps = 1
```
to work with fp16

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -274,7 +289,14 @@ def step(
             model_output = (sample - alpha_prod_t ** (0.5) * pred_original_sample) / beta_prod_t ** (0.5)
 
         # 6. compute "direction pointing to x_t" of formula (12) from https://arxiv.org/pdf/2010.02502.pdf
-        pred_sample_direction = (1 - alpha_prod_t_prev - std_dev_t**2) ** (0.5) * model_output
+        if self.prediction_type == "epsilon":
+            pred_sample_direction = (1 - alpha_prod_t_prev - std_dev_t**2) ** (0.5) * model_output
+
+            # 7. compute x_t without "random noise" of formula (12) from https://arxiv.org/pdf/2010.02502.pdf
+            prev_sample = alpha_prod_t_prev ** (0.5) * pred_original_sample + eps * pred_sample_direction
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:296]:
```suggestion
            prev_sample = alpha_prod_t_prev ** (0.5) * pred_original_sample + pred_sample_direction
```

**Review by patrickvonplaten — APPROVED** [review]
PR looks good, let's just make sure epsilon does not have the wrong dtype and device

**Review by pcuenca — APPROVED** [review]
LGTM! Same comments as @patrickvonplaten regarding setting `eps` seemingly unnecessarily.

We'll need `sample` prediction type for Euler to support k-upscaler, but we can add it later.

**Review by patrickvonplaten — APPROVED** [review]
Nice!

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 4]
```diff
@@ -229,7 +232,16 @@ def step(
...
         # 1. compute predicted original sample (x_0) from sigma-scaled predicted noise
-        pred_original_sample = sample - sigma_hat * model_output
+        if self.prediction_type == "epsilon":
+            pred_original_sample = sample - sigma_hat * model_output
+        elif self.prediction_type == "v_prediction":
+            # * c_out + input * c_skip
+            pred_original_sample = model_output * (-sigma / (sigma**2 + 1) ** 0.5) + (sample / (sigma**2 + 1))
+        else:
+            raise ValueError(
+                f"prediction_type given as {self.prediction_type} must be one of `epsilon`, `sample`, or"
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:242]:
```suggestion
                f"prediction_type given as {self.prediction_type} must be one of `epsilon`, or"
```

`sample` not yet supported here (but it will be) :)

**Review by pcuenca — COMMENTED** [review]
Cool!
